pkg/password/password_policies.go

```diff
@@ -0,0 +1,167 @@
+package password
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"unicode/utf8"
+)
+
+var defaultSpecialCharacters = "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"
+
+type Validator interface {
+	Validate(str string) error
+}
+
+type Policies struct {
+	minCharacters            int
+	minLowerCaseCharacters   int
+	minUpperCaseCharacters   int
+	minDigits                int
+	minSpecialCharacters     int
+	allowedSpecialCharacters string
+	digits                   *regexp.Regexp
+	specialCharacters        *regexp.Regexp
+}
+
+func NewPasswordPolicies(minCharacters, minLowerCaseCharacters, minUpperCaseCharacters, minDigits, minSpecialCharacters int,
+	allowedSpecialCharacters string) (Validator, error) {
+
+	digits := regexp.MustCompile("[0-9]")
+	if len(allowedSpecialCharacters) > 0 {
+		defaultSpecialCharacters = allowedSpecialCharacters
+	}
+	specialCharacters, err := regexp.Compile("[" + regexp.QuoteMeta(defaultSpecialCharacters) + "]")
+	if err != nil {
+		return nil, err
+	}
+	return &Policies{
+		minCharacters:            minCharacters,
+		minLowerCaseCharacters:   minLowerCaseCharacters,
+		minUpperCaseCharacters:   minUpperCaseCharacters,
+		minDigits:                minDigits,
+		minSpecialCharacters:     minSpecialCharacters,
+		allowedSpecialCharacters: allowedSpecialCharacters,
+		digits:                   digits,
+		specialCharacters:        specialCharacters,
+	}, nil
+}
+
+func NewDefaultPasswordPolicies() (Validator, error) {
+	return NewPasswordPolicies(0, 0, 0, 0, 0, "")
+}
+
+func (s Policies) Validate(str string) error {
+	var allErr error
+	if !utf8.ValidString(str) {
+		return fmt.Errorf("the password contains invalid characters")
+	}
+	err := s.validateCharacters(str)
+	if err != nil {
+		allErr = errors.Join(allErr, err)
+	}
+	err = s.validateLowerCase(str)
+	if err != nil {
+		allErr = errors.Join(allErr, err)
+	}
+	err = s.validateUpperCase(str)
+	if err != nil {
+		allErr = errors.Join(allErr, err)
+	}
+	err = s.validateDigits(str)
+	if err != nil {
+		allErr = errors.Join(allErr, err)
+	}
+	err = s.validateSpecialCharacters(str)
+	if err != nil {
+		allErr = errors.Join(allErr, err)
+	}
+	if allErr != nil {
+		return allErr
+	}
+	return nil
+}
+
+func (s Policies) validateCharacters(str string) error {
+	if s.count(str) < s.minCharacters {
+		if s.minCharacters == 1 {
+			return fmt.Errorf("at least one character is required")
+		}
```

Comment:
No need for this if case imo. The error message below is sufficient (also for other validate funcs)